### PR TITLE
Added keep-logs option to clean command

### PR DESCRIPTION
**Pull request by lorenzgerber**
For reclaiming space in /work while keeping all log information, we implemented an additional parameter for the clean command: nextflow clean -f -keepLogs
This deletes all temp files except these that match .commad.* or .exitcode. All directories are retained.

### CODE REVIEW

File: src/main/groovy/nextflow/cli/CmdClean.groovy [hunk 1]
```diff
@@ -228,4 +237,50 @@ class CmdClean extends CmdBase implements CacheBase {
...
+        Files.walkFileTree(folder, new FileVisitor<Path>() {
+
+            @Override
+            FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs) throws IOException {
+                result ? FileVisitResult.CONTINUE : FileVisitResult.TERMINATE
+            }
+
+            @Override
+            FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
+                if( !(file.getFileName() ==~ /^\.command.*/  || file.getFileName() ==~ /^\.exitcode/)){
```
Comment (pditommaso) [inline src/main/groovy/nextflow/cli/CmdClean.groovy]: Pattern matching is vet expensive for this operation, let's use instead:
```
file.name.startsWith('.command.') || file.name == '.exitcode'
```

File: src/main/groovy/nextflow/cli/CmdClean.groovy [hunk 2]
```diff
@@ -66,6 +66,9 @@ class CmdClean extends CmdBase implements CacheBase {
     @Parameter(names='-but', description = 'Clean up all runs except the specified one')
     String but
 
+    @Parameter(names='-keepLogs', description = 'Removes only temp files but keeps the "dot" files')
+    boolean keepLogs
```
Comment (pditommaso) [inline src/main/groovy/nextflow/cli/CmdClean.groovy]: For consistency I would propose to rename it to `-keep-logs`, maybe adding a `-k` shortcut. The label would be better as `Removes only temporary files but keeps execution logs and metadata`

**Comment by lorenzgerber** [comment]
'keep logs' functionality is now integrated in the `deleteFolder` method.

File: src/main/groovy/nextflow/cli/CmdClean.groovy [hunk 3]
```diff
@@ -201,9 +209,19 @@ class CmdClean extends CmdBase implements CacheBase {
...
+                        if( !file.delete() ) {
+                            result = false
+                            if(!quiet) System.err.println "Failed to remove $file"
+                        }
+                    }
+                } else {
+                    if( !file.delete() ) {
+                        result = false
+                        if(!quiet) System.err.println "Failed to remove $file"
+                    }
```
Comment (pditommaso) [inline src/main/groovy/nextflow/cli/CmdClean.groovy]: Looks bad having two identical if/else branches, Refactor it as 

```
final canDelete = !keepLogs && ( keepLogs &&  !(file.name.startsWith('.command.')  || file.name == '.exitcode'))
if( canDelete && !file.delete() ) { 
.. 
}
```

Comment (pditommaso) [reply src/main/groovy/nextflow/cli/CmdClean.groovy]: Also pass  `keepLogs` as a `deleteFolder` parameter to simplify unit testing

**Comment by pditommaso** [comment]
CLA received. Happy to merge this!

**Comment by lorenzgerber** [comment]
You are of course right. Clean with flag -k(eep logs) should result in an intermediate state where the 'full clean' command can still be applied to.

**Comment by pditommaso** [comment]
Yes, I think it makes sense.

**Pull request merged**